assembly/src/assembler/mod.rs

```diff
@@ -1,74 +1,289 @@
-        self.in_debug_mode = in_debug_mode;
+    pub fn with_debug_mode(mut self, yes: bool) -> Self {
+        self.in_debug_mode = yes;
         self
     }
 
+    /// Set whether to allow phantom calls
+    pub fn with_phantom_calls(mut self, yes: bool) -> Self {
+        self.allow_phantom_calls = yes;
+        self
+    }
+
+    /// Add `module` to the module graph of the assembler
+    ///
+    /// The given module must be a library module, or an error will be returned
+    pub fn with_module(mut self, module: Box<ast::Module>) -> Result<Self, Report> {
+        self.add_module(module)?;
+
+        Ok(self)
+    }
+
+    /// Add `module` to the module graph of the assembler
+    ///
+    /// The given module must be a library module, or an error will be returned
+    pub fn add_module(&mut self, module: Box<ast::Module>) -> Result<(), Report> {
+        match module.kind() {
+            ModuleKind::Kernel if self.kernel_index().is_some() => {
+                Err(Report::new(AssemblyError::ConflictingKernels))
+            }
+            ModuleKind::Kernel => {
+                let (kernel_index, kernel) = self.compile_kernel_module(module)?;
+                self.module_graph.set_kernel(Some(kernel_index), kernel);
+
+                Ok(())
+            }
+            ModuleKind::Executable => {
+                Err(Report::msg("cannot call `add_module` with an executable module: you must provide it via `compile` instead"))
+            }
+            ModuleKind::Library => {
+                self.module_graph.add_module(module)?;
+
+                Ok(())
+            }
+        }
+    }
+
+    /// Parse `source` as a library module with path `path`, and add it to the module graph of the assembler
+    pub fn with_module_from_source(
+        mut self,
+        path: LibraryPath,
+        source: impl ToString,
+    ) -> Result<Self, Report> {
+        self.add_module_from_source(path, source)?;
+        Ok(self)
+    }
+
+    /// Parse `source` as a library module with path `path`, and add it to the module graph of the assembler
+    pub fn add_module_from_source(
+        &mut self,
+        path: LibraryPath,
+        source: impl ToString,
+    ) -> Result<(), Report> {
+        let kind = ModuleKind::Library;
+        let source = Arc::new(SourceFile::new(path.path(), source.to_string()));
+        let ast = ast::Module::parse(path, kind, source)?;
+        self.add_module(ast)
+    }
+
     /// Adds the library to provide modules for the compilation.
-    pub fn with_library<L>(mut self, library: &L) -> Result<Self, AssemblyError>
+    pub fn with_library<L>(mut self, library: &L) -> Result<Self, Report>
     where
         L: Library,
     {
-        self.module_provider.add_library(library)?;
+        self.add_library(library)?;
+
         Ok(self)
     }
 
+    /// Adds the library to provide modules for the compilation.
+    pub fn add_library<L>(&mut self, library: &L) -> Result<(), Report>
```

Comment:
Yeah I'd be down with that, it wasn't clear to me why we passed it by ref anyway, but I could see arguments being made for it, but I think it largely ends up being a "six one way, half a dozen of the other" situation.

Reply:
It does occur to me that the one downside of making it owned rather than a reference is that we lose the ability to give `L` the `?Sized` negative-bound, i.e. making it unsized. This means, you couldn't pass a `&dyn Library` here, and for that reason alone I think it makes sense to keep this by-ref (and I've added the `?Sized` bound since it wasn't there).